### PR TITLE
[SPARK-23064][DOCS][SS] Added documentation for stream-stream joins

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?
Added documentation for stream-stream joins

image

image

image

image

## How was this patch tested?

N/a

### CODE REVIEW

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -1089,6 +1098,224 @@ streamingDf.join(staticDf, "type", "right_join")  # right outer join with a stat
...
+
+# Join with event-time constraints
+impressionsWithWatermark.join(
+  clicksWithWatermark,
+  expr("""
+    clickAdId = impressionAdId AND
+    clickTime >= impressionTime AND
+    clickTime <= impressionTime + interval 1 hour
+    """
+  ))
```
Comment (felixcheung) [inline docs/structured-streaming-programming-guide.md]: this *should* just work for R, like this:

(I added withWatermark in 2.3)

```
impressions <- read.stream( ...
clicks <- read.stream( ...

# Apply watermarks on event-time columns
impressionsWithWatermark <- withWatermark(impressions, "impressionTime", "2 hours")
clicksWithWatermark <- withWatermark(clicks, "clickTime", "3 hours")

# Join with event-time constraints
impressionsWithWatermark.join(
   clicksWithWatermark,
   expr(
     "clickAdId = impressionAdId AND
       clickTime >= impressionTime AND
       clickTime <= impressionTime + interval 1 hour"
))
```

File: docs/structured-streaming-programming-guide.md [hunk 2]
```diff
@@ -1089,6 +1098,224 @@ streamingDf.join(staticDf, "type", "right_join")  # right outer join with a stat
...
+
+##### Outer Joins with Watermarking
+While the watermark + event-time constraints is optional for inner joins, for left and right outer
+joins they must be specified. This is because for generating the NULL results in outer join, the
+engine must know when an input row is not going to match with anything in future. Hence, the
+watermark + event-time constraints must be specified for generating correct results.
+
+However, note that the outer NULL results will be generated with a delay (depends on the specified
+watermark delay and the time range condition) because the engine has to wait for that long to ensure
+ there were no matches and there will be no more matches in future.
```
Comment (felixcheung) [inline docs/structured-streaming-programming-guide.md]: extra space?

**Review by felixcheung — APPROVED** [review]

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (tdas) [reply docs/structured-streaming-programming-guide.md]: Thank you!

Comment (tdas) [reply docs/structured-streaming-programming-guide.md]: Could you add tests for joins in R as well? :)

File: docs/structured-streaming-programming-guide.md [hunk 3]
```diff
@@ -2142,6 +2452,7 @@ write.stream(aggDF, "memory", outputMode = "complete", checkpointLocation = "pat
 
 **Talks**
 
-- Spark Summit 2017 Talk - [Easy, Scalable, Fault-tolerant Stream Processing with Structured Streaming in Apache Spark](https://spark-summit.org/2017/events/easy-scalable-fault-tolerant-stream-processing-with-structured-streaming-in-apache-spark/)
+- Spark Summit Europe 2017 Talks -
+  - [Easy, Scalable, Fault-tolerant Stream Processing with Structured Streaming in Apache Spark](https://spark-summit.org/2017/events/easy-scalable-fault-tolerant-stream-processing-with-structured-streaming-in-apache-spark/)
```
Comment (tdas) [inline docs/structured-streaming-programming-guide.md]: TODO: this link needs to be updated. blocked on some links not working on the spark summit website.

File: docs/structured-streaming-programming-guide.md [hunk 4]
```diff
@@ -1089,6 +1101,300 @@ streamingDf.join(staticDf, "type", "right_join")  # right outer join with a stat
...
+    <td style="vertical-align: middle;">Full Outer</td>
+    <td style="vertical-align: middle;">Not supported</td>
+  </tr>
+  <tr>
+    <td rowspan="4" style="vertical-align: middle;">Stream</td>
+    <td rowspan="4" style="vertical-align: middle;">Stream</td>
+    <td style="vertical-align: middle;">Inner</td>
+    <td style="vertical-align: middle;">
+      Supported, optionally specify watermark on both sides +
+      time constraints for state cleanup<
```
Comment (zsxwing) [inline docs/structured-streaming-programming-guide.md]: nit" remove `<`

**Review by zsxwing — CHANGES_REQUESTED** [review]
LGTM except some nits

**Review by zsxwing — APPROVED** [review]
LGTM

**Comment by zsxwing** [comment]
Merging to master and 2.3.

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (felixcheung) [reply docs/structured-streaming-programming-guide.md]: sure!

**Pull request closed without merging**